### ISSUE TITLE
The map and map50 output by YOLO is inconsistent with the map and map50 output by COCO Eval.

### ISSUE THREAD

**Issue by FreemanTang** [ultralytics/ultralytics#24637]
### Search before asking

- [x] I have searched the Ultralytics YOLO [issues](https://github.com/ultralytics/ultralytics/issues) and found no similar bug report.

### Ultralytics YOLO Component

Val

### Bug

@Laughing-q Hello, I have a few questions:
1. Is the map provided here the validation results generated by Ultralytics YOLO, or are they results obtained using the Coco API evaluation?
<img width="465" height="460" alt="Image" src="https://github.com/user-attachments/assets/e0ba4857-41f2-4b8b-9e81-b54524357d30" />

**2.  Does Ultralytics provide the results from the COCO API evaluation internally?（The main purpose is to know whether Ultralytics officially provides direct COCO API evaluation, which can output COCO evaluation indicators during training and verification for easy comparison.）**
3.  After running `metrics = model.val(data=args.data, save_json=True)`, I used pycocotools to perform a COCO evaluation and found that the mAP50-95 and mAP50 values from the built-in YOLO model differ from those in the COCO evaluation.
```
Ultralytics 8.4.50 🚀 Python-3.11.8 torch-2.7.1+cu128 CUDA:0 (NVIDIA GeForce RTX 3080 Ti, 12045MiB)
YOLO26s summary (fused): 122 layers, 9,465,567 parameters, 0 gradients, 20.5 GFLOPs
val: Fast image access ✅ (ping: 0.0±0.0 ms, read: 2922.7±490.2 MB/s, size: 942.9 KB)
val: Scanning ./AI_test/datasets/detect/data_20260514_142239/labels/val.cache... 323 images, 0 backgrounds, 0 corrupt: 100% ━━━━━━━━━━━━ 323/323 42.3Mit/s 0.0s
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100% ━━━━━━━━━━━━ 21/21 5.7it/s 3.7s
                   all        323        794      0.939      0.788      0.839      0.622

creating index...
index created!
Loading and preparing results...
DONE (t=0.00s)
creating index...
index created!
Running per image evaluation...
Evaluate annotation type *bbox*
DONE (t=0.27s).
Accumulating evaluation results...
DONE (t=0.04s).
 Average Precision  (AP) @[ IoU=0.50:0.95 | area=   all | maxDets=100 ] = 0.603
 Average Precision  (AP) @[ IoU=0.50      | area=   all | maxDets=100 ] = 0.836
 Average Precision  (AP) @[ IoU=0.75      | area=   all | maxDets=100 ] = 0.664
 Average Precision  (AP) @[ IoU=0.50:0.95 | area= small | maxDets=100 ] = 0.357
 Average Precision  (AP) @[ IoU=0.50:0.95 | area=medium | maxDets=100 ] = 0.607
 Average Precision  (AP) @[ IoU=0.50:0.95 | area= large | maxDets=100 ] = 0.611
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=   all | maxDets=  1 ] = 0.307
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=   all | maxDets= 10 ] = 0.641
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=   all | maxDets=100 ] = 0.657
 Average Recall     (AR) @[ IoU=0.50:0.95 | area= small | maxDets=100 ] = 0.447
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=medium | maxDets=100 ] = 0.655
 Average Recall     (AR) @[ IoU=0.50:0.95 | area= large | maxDets=100 ] = 0.663
```

Since private datasets cannot be made public, the Ultralytics team could use `coco8.yaml` as an example for testing. I previously noticed that models trained on `coco8.yaml` seem to produce different results for YOLO’s built-in `map` and `map50` metrics compared to those evaluated using COCO’s `map` and `map50`. I haven’t been able to identify the cause of this issue yet, so I look forward to your response!

### Environment

Ultralytics 8.4.50 🚀 Python-3.11.8 torch-2.7.1+cu128 CUDA:0 (NVIDIA GeForce RTX 3080 Ti, 12045MiB)
OS  Ubuntu 20.04.6 LTS
Python 3.11.8

### Minimal Reproducible Example

[predictions.json](https://github.com/user-attachments/files/28372157/predictions.json)
[val.json](https://github.com/user-attachments/files/28372156/val.json)

predictions.json is the prediction result, val.json is the real result。

### Are you willing to submit a PR?

- [ ] Yes I'd like to help by submitting a PR!

**Comment by UltralyticsAssistant** [comment]
👋 Hello @FreemanTang, thank you for your interest in Ultralytics 🚀! This is an automated response to help get your issue routed quickly, and an Ultralytics engineer will also assist you soon.
We recommend a visit to the [Docs](https://docs.ultralytics.com/) for new users where you can find many [Python](https://docs.ultralytics.com/usage/python) and [CLI](https://docs.ultralytics.com/usage/cli) usage examples and where many of the most common questions may already be answered.
Since this appears to be a 🐛 bug report related to validation metrics, thank you for already including a reproducible setup and shared files. If possible, please also add a small **minimum reproducible example** following the [minimum reproducible example](https://docs.ultralytics.com/help/minimum-reproducible-example/) guide that shows the exact validation command, the exact COCO evaluation script, and any relevant arguments used in both runs. This will help us compare the two evaluation paths more quickly 🔍
If this is a custom training ❓ question, please provide as much information as possible, including dataset image examples and training logs, and verify you are following our [Tips for Best Training Results](https://docs.ultralytics.com/guides/model-training-tips).
Join the Ultralytics community where it suits you best. For real-time chat, head to [Discord](https://discord.com/invite/ultralytics) 🎧. Prefer in-depth discussions? Check out [Discourse](https://community.ultralytics.com/). Or dive into threads on our [Subreddit](https://www.reddit.com/r/Ultralytics/) to share knowledge with the community.
## Upgrade
Upgrade to the latest `ultralytics` package including all [requirements](https://github.com/ultralytics/ultralytics/blob/main/pyproject.toml) in a [**Python>=3.8**](https://www.python.org/) environment with [**PyTorch>=1.8**](https://pytorch.org/get-started/locally/) to verify your issue is not already resolved in the latest version:
```bash
pip install -U ultralytics
```
## Environments
YOLO may be run in any of the following up-to-date verified environments (with all dependencies including [CUDA](https://developer.nvidia.com/cuda)/[CUDNN](https://developer.nvidia.com/cudnn), [Python](https://www.python.org/) and [PyTorch](https://pytorch.org/) preinstalled):
- **Notebooks** with free GPU: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolo26"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **Google Cloud** Deep Learning VM. See [GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial)
- **Amazon** Deep Learning AMI. See [AWS Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial)
- **Docker Image**. See [Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial) <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>
## Status
<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml/badge.svg" alt="Ultralytics CI"></a>
If this badge is green, all [Ultralytics CI](https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule) tests are currently passing. CI tests verify correct operation of all YOLO [Modes](https://docs.ultralytics.com/modes) and [Tasks](https://docs.ultralytics.com/tasks) on macOS, Windows, and Ubuntu every 24 hours and on every commit.

**Comment by Y-T-G** [comment]
Hello.

1. The mAP reported on Ultralytics Docs is obtained uaing COCOEval which runs automatically if you run validation with `data=coco.yaml`.
2. Ultralytics only calculates COCOEval mAP for validation on `coco.yaml` and `coco-pose.yaml`. Otherwise, Ultralytics uses it's own native implementation to calculate mAP, which is close to COCOEval but not identical.
3. The discrepancy is known and expected. The internal mAP calculation isn't identical to COCOEval.

**Comment by FreemanTang** [comment]
> Hello.
> 
> 1. The mAP reported on Ultralytics Docs is obtained uaing COCOEval which runs automatically if you run validation with `data=coco.yaml`.
> 2. Ultralytics only calculates COCOEval mAP for validation on `coco.yaml` and `coco-pose.yaml`. Otherwise, Ultralytics uses it's own native implementation to calculate mAP, which is close to COCOEval but not identical.
> 3. The discrepancy is known and expected. The internal mAP calculation isn't identical to COCOEval.

@Y-T-G Thanks for your reply! Please tell me, if I train my own training set, how can I calculate COCOEval mAP during verification like `coco.yaml` and `coco-pose.yaml`?

**Comment by Y-T-G** [comment]
You can use `save_json=True` to save the predictions and then run COCOEval with it. You also need your original labels in COCO JSON format for COCOEval.

**Comment by FreemanTang** [comment]
> You can use `save_json=True` to save the predictions and then run COCOEval with it. You also need your original labels in COCO JSON format for COCOEval.

@Y-T-G  Thank you for your reply. I did what you said here, but there is a big difference between the map calculated by COCOEval and the map calculated by Ultralytics. I’m not sure where the problem lies. Does Ultralytics officially provide code examples for COCOEval?

> [predictions.json](https://github.com/user-attachments/files/28372157/predictions.json) [val.json](https://github.com/user-attachments/files/28372156/val.json)
> 
> predictions.json is the prediction result, val.json is the real result。
> **2. Does Ultralytics provide the results from the COCO API evaluation internally?（The main purpose is to know whether Ultralytics officially provides direct COCO API evaluation, which can output COCO evaluation indicators during training and verification for easy comparison.）** 3. After running `metrics = model.val(data=args.data, save_json=True)`, I used pycocotools to perform a COCO evaluation and found that the mAP50-95 and mAP50 values from the built-in YOLO model differ from those in the COCO evaluation.
> 
> ```
> Ultralytics 8.4.50 🚀 Python-3.11.8 torch-2.7.1+cu128 CUDA:0 (NVIDIA GeForce RTX 3080 Ti, 12045MiB)
> YOLO26s summary (fused): 122 layers, 9,465,567 parameters, 0 gradients, 20.5 GFLOPs
> val: Fast image access ✅ (ping: 0.0±0.0 ms, read: 2922.7±490.2 MB/s, size: 942.9 KB)
> val: Scanning ./AI_test/datasets/detect/data_20260514_142239/labels/val.cache... 323 images, 0 backgrounds, 0 corrupt: 100% ━━━━━━━━━━━━ 323/323 42.3Mit/s 0.0s
>                  Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100% ━━━━━━━━━━━━ 21/21 5.7it/s 3.7s
>                    all        323        794      0.939      0.788      0.839      0.622
> 
> creating index...
> index created!
> Loading and preparing results...
> DONE (t=0.00s)
> creating index...
> index created!
> Running per image evaluation...
> Evaluate annotation type *bbox*
> DONE (t=0.27s).
> Accumulating evaluation results...
> DONE (t=0.04s).
>  Average Precision  (AP) @[ IoU=0.50:0.95 | area=   all | maxDets=100 ] = 0.603
>  Average Precision  (AP) @[ IoU=0.50      | area=   all | maxDets=100 ] = 0.836
>  Average Precision  (AP) @[ IoU=0.75      | area=   all | maxDets=100 ] = 0.664
>  Average Precision  (AP) @[ IoU=0.50:0.95 | area= small | maxDets=100 ] = 0.357
>  Average Precision  (AP) @[ IoU=0.50:0.95 | area=medium | maxDets=100 ] = 0.607
>  Average Precision  (AP) @[ IoU=0.50:0.95 | area= large | maxDets=100 ] = 0.611
>  Average Recall     (AR) @[ IoU=0.50:0.95 | area=   all | maxDets=  1 ] = 0.307
>  Average Recall     (AR) @[ IoU=0.50:0.95 | area=   all | maxDets= 10 ] = 0.641
>  Average Recall     (AR) @[ IoU=0.50:0.95 | area=   all | maxDets=100 ] = 0.657
>  Average Recall     (AR) @[ IoU=0.50:0.95 | area= small | maxDets=100 ] = 0.447
>  Average Recall     (AR) @[ IoU=0.50:0.95 | area=medium | maxDets=100 ] = 0.655
>  Average Recall     (AR) @[ IoU=0.50:0.95 | area= large | maxDets=100 ] = 0.663
> ```
> 
> Since private datasets cannot be made public, the Ultralytics team could use `coco8.yaml` as an example for testing. I previously noticed that models trained on `coco8.yaml` seem to produce different results for YOLO’s built-in `map` and `map50` metrics compared to those evaluated using COCO’s `map` and `map50`. I haven’t been able to identify the cause of this issue yet, so I look forward to your response!
> 
> ### Environment
> Ultralytics 8.4.50 🚀 Python-3.11.8 torch-2.7.1+cu128 CUDA:0 (NVIDIA GeForce RTX 3080 Ti, 12045MiB) OS Ubuntu 20.04.6 LTS Python 3.11.8
> 
> ### Minimal Reproducible Example
> [predictions.json](https://github.com/user-attachments/files/28372157/predictions.json) [val.json](https://github.com/user-attachments/files/28372156/val.json)
> 
> predictions.json is the prediction result, val.json is the real result。

**Comment by Y-T-G** [comment]
The mAP calculated by Ultralytics would be different from COCOEval because there are differences in calculation which you can read here:

https://github.com/ultralytics/ultralytics/issues/17270#issuecomment-2503639995

**Comment by FreemanTang** [comment]
> The mAP calculated by Ultralytics would be different from COCOEval because there are differences in calculation which you can read here:
> 
> [#17270 (comment)](https://github.com/ultralytics/ultralytics/issues/17270#issuecomment-2503639995)

https://github.com/ultralytics/ultralytics/issues/17270#issuecomment-2503639995

@Y-T-G Thank you for your reply! I took a look at https://github.com/ultralytics/ultralytics/issues/17270, and I have a question I’d like to ask you: After you identified the likely cause, which script did you use to calculate the mAP for testing? Could you please share it so I can test it on my own dataset and model?

<img width="909" height="799" alt="Image" src="https://github.com/user-attachments/assets/63bc9988-1746-45f1-82d5-0de812409e1e" />

**Comment by Y-T-G** [comment]
It's a modified COCOEval script. I don't have it now.

**Comment by FreemanTang** [comment]
<img width="909" height="799" alt="Image" src="https://github.com/user-attachments/assets/0f11ec4f-e5e1-4c60-9447-64af839ff6e8" />

> It's a modified COCOEval script. I don't have it now.

@Y-T-G  Thanks for your reply! Logically speaking, this is your test in 2024. Does it mean that Ultralytics 8.4.50 has fixed this problem?
When using model.val(), besides save_json=true, are there any parameters that need to be set? For example, should max_det be set to 100?

Also, I found that Ultralytics provides an [ultralytics.models.yolo.detect.val.DetectionValidator.coco_evaluate](https://docs.ultralytics.com/reference/models/yolo/detect/val#ultralytics.models.yolo.detect.val.DetectionValidator.coco_evaluate) method. Is this coco eval accurate? Are there any specific usage examples?

<img width="921" height="375" alt="Image" src="https://github.com/user-attachments/assets/cd0c2a72-f593-4a37-be94-8956e876d77d" />

**Comment by Y-T-G** [comment]
Ultralytics automatically uses COCOEval when evaluating on MS COCO.

And the bug in the issue is fixed. So mAP would be close to COCOEval. But it's not the same as COCOEval. COCOEval handles labels with crowded tag differently. Ultralytics doesn't have that handling

**Comment by FreemanTang** [comment]
> Ultralytics automatically uses COCOEval when evaluating on MS COCO.
> 
> And the bug in the issue is fixed. So mAP would be close to COCOEval. But it's not the same as COCOEval. COCOEval handles labels with crowded tag differently. Ultralytics doesn't have that handling

@Y-T-G Thank you for your reply! I’d like to ask for your advice: based on your experience, how would you recommend I go about comparing the mAP values of different models (such as PPYOLOE and YOLO26)? PPYOLOE outputs COCO mAP, while YOLO26 outputs YOLO mAP. 

If the outputs aren’t standardized to COCO mAP, the comparison won’t be very rigorous.

**Comment by Y-T-G** [comment]
Ultralytics will output COCOEval mAP if you validate on `data=coco.yaml`. Ultralytics always uses COCOEval mAP for the mAP shown in documentation and comparison tables.

If you're using a custom dataset, then you need to use `save_json=True` to save your predictions and then run COCOEval manually with your data converted to COCO JSON format
